examples/kmeans.cpp

```diff
@@ -0,0 +1,148 @@
+    int maxIter = std::stoi(husky::Context::get_param("iter"));
+
+    husky::lib::Aggregator<int> dataCountAgg;
+    auto& ac = husky::lib::AggregatorFactory::get_channel();
+
+    // 1. Create and globalize point objects
+    auto& point_list = husky::ObjListStore::create_objlist<Point>();
+    auto parse_row = [&](boost::string_ref& chunk) {
+        if (chunk.size() == 0)
+            return;
+        boost::char_separator<char> sep(" \t");
+        boost::tokenizer<boost::char_separator<char>> tok(chunk, sep);
+        boost::tokenizer<boost::char_separator<char>>::iterator it = tok.begin();
+        int id = stoi(*it++);
+        featureVec fV(dim);
+        for (int idx = 0; idx < dim; ++idx) {
+            float tmp = stod(*it++);
+            fV[idx] = tmp;
+        }
+
+        dataCountAgg.update(1);
+        point_list.add_object(Point(id, fV));
+    };
+    load(infmt, {&ac}, parse_row);
+    husky::globalize(point_list);
+
+    int totDataPoints = dataCountAgg.get_value();
+    husky::base::log_info("point_list size " + std::to_string(totDataPoints));
+
+    // 2. train
+    std::vector<featureVec> init_center(3, featureVec(dim, 0.));
+    init_center[0][0] = 0.;
+    init_center[0][1] = 1.;
+    init_center[1][0] = -1.;
+    init_center[1][1] = -1.;
+    init_center[2][0] = 1.;
+    init_center[2][1] = 1.;
+
+    auto kmeansOp = husky::lib::ml::Kmeans<Point, featureVec>(
+        point_list,                                // obj_list
+        [](const Point& p){ return p.features;},   // lambda to extraxt feature from obj
+        euclidean_dist,                            // lambda to calculate distance between 2 features
+        3,                                         // no. of centers to use, OR put init_center here
+        maxIter,                                   // max iteration
+        husky::lib::ml::KmeansOpts::kInitKmeansPP  // (optional arg) init method, kInitKmeansPP OR kInitSimple,
+                                                   // irrelevant if init_center provided
+    );
+
+    kmeansOp.run();
+
+    // 3. output
+    if (husky::Context::get_global_tid() == 0) {
+        auto clusterCenter = kmeansOp.clusterCenter;
+        std::ostringstream oss;
+        oss << std::scientific;
+        oss.precision(17);
+        oss << "kmeans: " << std::endl;
+        for (int i = 0; i < clusterCenter.size(); ++i) {
+            oss << i << " ";
+            for (auto aVal : clusterCenter[i]) {
+                oss << aVal << " ";
+            }
+            oss << std::endl;
+        }
+        husky::base::log_info(oss.str());
+    }
+}
+
+int main(int argc, char** argv) {
+    // Input:
+    // A txt with each line containing 1 row of the data as
+    // rowID rowElem1 rowElem2 rowElem3 ...rowElemN
+    std::vector<std::string> args;
+    args.push_back("hdfs_namenode");
+    args.push_back("hdfs_namenode_port");
+    args.push_back("input");  // path to input file eg. hdfs:///user/ylchan/AffMat_T2/merge
```

Comment:
Overall, I think kmeans.cpp is too long. We would like to put as many code in lib as possible. You may improve your code as follows:

1. Accept only LIBSVM format (we can allow users to define their parser later on)
2. Return an obj_list that contains information of clusters

You may refer to Spark Kmeans. I think they exhibit a very simple API. 
http://spark.apache.org/docs/latest/ml-clustering.html